sel/src/Sel/Hashing/Generic.hs

```diff
@@ -0,0 +1,171 @@
+{-# LANGUAGE DerivingStrategies #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE InstanceSigs #-}
+
+-- |
+--
+-- Module: Sel.Hashing.Generic
+-- Description: Fingerprint hashing with the BLAKE2b algorithm
+-- Copyright: (C) Hécate Moonlight 2022
+-- License: BSD-3-Clause
+-- Maintainer: The Haskell Cryptography Group
+-- Portability: GHC only
+module Sel.Hashing.Generic
+  ( -- * Introduction
+    -- $introduction
+
+    -- * Operations
+    HashKey
+  , newHashKey
+  , Hash
+  , hashByteString
+  , hashToText
+  , hashToByteString
+  , hashToBinary
+  )
+where
+
+import Control.Monad (void)
+import Data.ByteString (ByteString)
+import qualified Data.ByteString.Base16 as Base16
+import Data.ByteString.Unsafe (unsafePackMallocCStringLen, unsafeUseAsCStringLen)
+import Data.Text (Text)
+import Data.Text.Display
+import qualified Data.Text.Lazy.Builder as Builder
+import Foreign (ForeignPtr, Ptr)
+import qualified Foreign
+import Foreign.C (CChar, CSize, CUChar)
+import Foreign.Storable
+import GHC.IO.Handle.Text (memcpy)
+import LibSodium.Bindings.GenericHashing (cryptoGenericHash, cryptoGenericHashBytes, cryptoGenericHashKeyBytes, cryptoGenericHashKeyGen)
+import System.IO.Unsafe (unsafeDupablePerformIO)
+
+-- $introduction
+--
+-- This API computes a fixed-length fingerprint for an arbitrarily long message.
+-- It is backed by the BLAKE2b algorithm.
+--
+-- Sample use cases:
+--
+--   * File integrity checking
+--   * Creating unique identifiers to index arbitrarily long data
+--
+-- __⚠️ Do not use this module to hash passwords! ⚠️__
+--
+-- If you need to deviate from the defaults enforced by this module,
+-- please use the underlying bindings at "LibSodium.Bindings.GenericHashing".
+
+-- |
+--
+-- @since 0.0.1.0
+newtype HashKey = HashKey (ForeignPtr CUChar)
+  deriving newtype
+    ( Eq
+      -- ^ @since 0.0.1.0
+    , Ord
+      -- ^ @since 0.0.1.0
+    , Show
+      -- ^ @since 0.0.1.0
+    )
```

Comment:
@kozross You do not recommend the derivations in principle, or as they stand now and should be properly implemented?

Reply:
I don't recommend them in principle, because they're almost _never_ what you want. In this specific case, they're not what you want instead.

Basically, implement these to do what you actually want here, not compare pointers.